vdr/vdr.go

```diff
@@ -132,32 +132,98 @@ func (r VDR) Resolve(id did.DID, metadata *types.ResolveMetadata) (*did.Document
-	resolverMetada := &types.ResolveMetadata{
+	resolverMetadata := &types.ResolveMetadata{
 		Hash:             &current,
-		AllowDeactivated: false,
+		AllowDeactivated: true,
 	}
-	currentDIDdocument, meta, err := r.store.Resolve(id, resolverMetada)
+	currentDIDDocument, meta, err := r.store.Resolve(id, resolverMetadata)
 	if err != nil {
 		return err
 	}
+	if isDeactivated(currentDIDDocument) {
+		return types.ErrDeactivated
+	}
 	payload, err := json.Marshal(next)
 	if err != nil {
 		return err
 	}
 
+	controllers, err := r.resolveControllers([]did.Document{*currentDIDDocument})
+	if err != nil {
+		return fmt.Errorf("error while finding controllers for document: %w", err)
+	}
+	if len(controllers) == 0 {
+		return fmt.Errorf("could not find any controllers for document")
+	}
+
 	// TODO: look into the controller of the did for a signing key
-	keyID := currentDIDdocument.Authentication[0].ID.String()
+	keyID := controllers[0].Authentication[0].ID.String()
 	_, err = r.network.CreateTransaction(didDocumentType, payload, keyID, nil, time.Now(), dag.TimelineIDField(meta.TimelineID), dag.TimelineVersionField(meta.Version+1))
-
 	if err == nil {
 		logging.Log().Infof("DID Document updated: %s", id)
 	}
 
 	return err
 }
 
+func isDeactivated(document *did.Document) bool {
+	return len(document.Controller) == 0 && len(document.Authentication) == 0
+}
+
 // Deactivate updates the DID Document so it can no longer be updated
-func (r *VDR) Deactivate(DID did.DID, current hash.SHA256Hash) {
-	logging.Log().Debugf("Deactivating DID Document: %s", DID)
-	panic("implement me")
+func (r *VDR) Deactivate(id did.DID, currentHash hash.SHA256Hash) error {
+	// A deactivated document is the original document stripped from a controller and keys.
+	// So, apart from the services, it is practically an empty document. Should we even keep the services?
+	emptyDoc := did.Document{ID: id}
+	return r.Update(id, currentHash, emptyDoc, nil)
+}
+
+// resolveControllers accepts a list of documents and finds their controllers
+// The resulting list are documents who control themselves
+func (r *VDR) resolveControllers(input []did.Document) ([]did.Document, error) {
+	// end of the chain
+	if len(input) == 0 {
+		return input, nil
+	}
+
+	var leaves []did.Document
+	var refsToResolve []did.DID
+	var nodes []did.Document
+
+	// for each input document, find its controllers or add the doc itself if its controls itself
+	for _, doc := range input {
+		if len(doc.Controller) == 0 && len(doc.Authentication) > 0 {
+			// no controller -> doc is its own controller
+			leaves = append(leaves, doc)
+			continue
+		}
+		for _, ctrlDID := range doc.Controller {
+			if doc.ID.Equals(ctrlDID) {
+				// doc is its own controller
+				leaves = append(leaves, doc)
+			} else {
+				// add did to be resolved later
+				refsToResolve = append(refsToResolve, ctrlDID)
+			}
+		}
+	}
+	// resolve all unresolved docs
+	// TODO: check for loops or already resolved docs?
```

Comment:
these are 2 TODO's;

- check for already resolved docs is simple to do, just loop the input array?
- check for loops is important; and should be clear from the spec/RFC

Reply:
Added the issue of recursive controllers to: https://github.com/nuts-foundation/nuts-specification/issues/39